language/diem-framework/modules/DiemId.move

```diff
@@ -0,0 +1,144 @@
+        removed: bool,
+        /// Diem ID Domain string of the account
+        domain: DiemIdDomain,
+        /// On-chain account address
+        address: address,
+    }
+
+    const MAX_U64: u128 = 18446744073709551615;
+
+    // Error codes
+    /// DiemIdDomains resource is not or already published.
+    const EDIEMIDDOMAIN: u64 = 0;
+    /// DiemIdDomainManager resource is not or already published.
+    const EDIEMIDDOMAINMANAGER: u64 = 1;
+
+    /// Publish a `DiemIdDomains` resource under `created` with an empty `domains`.
+    /// Before sending or receiving any payments using Diem IDs, the Treasury Compliance account must send
+    /// a transaction that invokes `add_domain_id` to set the `domains` field with a valid domain
+    public fun publish_diem_id_domains(
+        created: &signer,
+    ) {
+        Roles::assert_parent_vasp_role(created);
+        assert(
+            !exists<DiemIdDomains>(Signer::address_of(created)),
+            Errors::already_published(EDIEMIDDOMAIN)
+        );
+        move_to(created, DiemIdDomains {
+            domains: Vector::empty(),
+        })
+    }
+    spec fun publish_diem_id_domains {
+        include Roles::AbortsIfNotParentVasp{account: created};
+        aborts_if has_diem_id_domains(Signer::spec_address_of(created)) with Errors::ALREADY_PUBLISHED;
+    }
+
+    public fun has_diem_id_domains(addr: address): bool {
+        exists<DiemIdDomains>(addr)
+    }
+
+    /// Publish a `DiemIdDomainManager` resource under `created` with an empty `diem_id_domain_events`.
+    /// When Treasury Compliance account sends a transaction that invokes `update_diem_id_domain`,
+    /// a `DiemIdDomainEvent` is emitted and added to `diem_id_domain_events`.
+    public fun publish_diem_id_domain_manager(
+        created: &signer,
+    ) {
+        Roles::assert_treasury_compliance(created);
+        assert(
+            !exists<DiemIdDomainManager>(Signer::address_of(created)),
+            Errors::already_published(EDIEMIDDOMAINMANAGER)
+        );
+        move_to(
+            created,
+            DiemIdDomainManager {
+                diem_id_domain_events: Event::new_event_handle<DiemIdDomainEvent>(created),
+            }
+        );
+    }
+
+    spec define spec_has_diem_id_domain_manager(addr: address): bool {
+        exists<DiemIdDomainManager>(addr)
+    }
+
+    spec fun publish_diem_id_domain_manager {
+        include Roles::AbortsIfNotTreasuryCompliance{account: created};
+        aborts_if spec_has_diem_id_domain_manager(Signer::spec_address_of(created)) with Errors::ALREADY_PUBLISHED;
+    }
+
+    public fun update_diem_id_domain(
+        tc_account: &signer,
+        to_update_address: address,
+        domain: vector<u8>,
+        is_remove: bool
+    ) acquires DiemIdDomainManager, DiemIdDomains {
+        Roles::assert_treasury_compliance(tc_account);
+        let account_domains = borrow_global_mut<DiemIdDomains>(to_update_address);
```

Comment:
This can fail if `to_update_address` does not have a `DiemIdDomains` resource--would you mind aborting with a descriptive error code in this case?

In the DF, we try to enforce a policy of "no uncovered aborts"--that is, every failing transaction should abort with a documented error code.